Add support for use of existing VPC with Cloudformation

webvictim opened the pull request:
Fixes #1956

webvictim committed:
Ignore vim swap files

webvictim committed:
Ignore vim swap files

webvictim committed:
Add support for existing VPC

webvictim committed:
Add support for existing VPC

webvictim committed:
Remove old VPC code from YAML

webvictim committed:
Remove old VPC code from YAML

gravitational-jenkins commented:
Can one of the admins verify this patch?

klizhentas commented on assets/marketplace/oss-existingvpc.yaml:
```diff
@@ -0,0 +1,1064 @@
+    Type: AWS::EC2::KeyPair::KeyName
+
+  InstanceType:
+    Description: WebServer EC2 instance type    
```
WebServer -> Teleport

klizhentas commented on assets/marketplace/oss-existingvpc.yaml:
```diff
@@ -0,0 +1,1064 @@
+    Type: String
+
+  # Any existing VPC used must have both DNS support and DNS hostnames enabled
+  VPC:
```
missing description

klizhentas commented on assets/marketplace/oss-existingvpc.yaml:
```diff
@@ -0,0 +1,1064 @@
+Resources:
+  # Internet gateway is a routing target for 
+  # internet routeable target
+  InternetGateway:
```
no need to create a gateway, usually VPC should have one already

klizhentas commented on assets/marketplace/oss-existingvpc.yaml:
```diff
@@ -0,0 +1,1064 @@
+  # route to use internet gateway,
+  # subnets using this table are public subnets
+  # from AWS perspective
+  PublicRoutingTable:
```
same here

klizhentas commented on assets/marketplace/oss-existingvpc.yaml:
```diff
@@ -0,0 +1,1064 @@
+      Domain: vpc
+
+  # NAT gateway in the first AZ
+  NATA:
```
same for nat gateway - should require two nat gateways - one per AZ

webvictim replied:
I don't think CloudFormation allows you to pass in the ID of an existing NAT Gateway to this function, it's expecting to create it itself and associate it with the ProxySubnet.

We could maybe ask for the IDs of ProxySubnetA and ProxySubnetB to be passed in as parameters rather than the CIDR block? I'm not sure how we validate that they are public rather than private subnets if we don't create them, though.

klizhentas commented on assets/marketplace/oss-existingvpc.yaml:
```diff
@@ -0,0 +1,1064 @@
+      Protocol: TCP
+      VpcId: {Ref: VPC}
+
+  ProxyLaunchConfig: 
```
would be cool if we found a way to re-use ASG configs, otherwise we will end up with two copies of the configuration

webvictim replied:
I think the problem here is the restriction of us only being able to use one big YAML file for the marketplace - so no includes. We could potentially put the ProxyLaunchConfig stuff in a separate file, then write some kind of script to substitute a placeholder value in each YAML file with the included config?

klizhentas reviewed (changes requested):
Needs some changes

webvictim committed:
Integrate README updates from rjones

webvictim committed:
Integrate README updates from rjones

webvictim committed:
README tidying

webvictim committed:
README tidying

webvictim committed:
Add latest changes splitting into VPC creation and ASG creation

webvictim committed:
Fix merge conflicts

webvictim committed:
JENKINS_TEST_STACK_NAME should no longer be a constant

webvictim committed:
Update stack names and add extra output description command

webvictim committed:
Merge branch 'master' into gus/cf-existing-vpc-support

webvictim committed:
Disable post steps in Jenkinsfile for debugging

webvictim committed:
Fixed typo in stack command

webvictim committed:
Make should be silent when outputting JSON

webvictim committed:
Update parameters path

webvictim committed:
Typo

webvictim committed:
Make parameters available for previous step

webvictim committed:
Restore Jenkinsfile

webvictim committed:
Merge branch 'master' into gus/cf-existing-vpc-support

webvictim committed:
CloudFormation should also output S3 bucket ID

webvictim committed:
Try to clean up S3 bucket when we exit

webvictim committed:
Add s3-delete-all-object-versions script

webvictim committed:
Don't run post

webvictim committed:
Turn up start time to give us more time to debug

webvictim committed:
Add dynamic domain name generation to avoid rate limits

webvictim committed:
Formatting

webvictim committed:
Variables aren't groovy

webvictim committed:
Limit random output

webvictim committed:
Limit random output

webvictim committed:
Disable Packer for now and limit output range

webvictim committed:
Leave existing stack in place

webvictim committed:
Rename file

webvictim committed:
Write string differently

webvictim committed:
Re-enable Packer

webvictim committed:
Make domain name shorter so it doesn't exceed Letsencrypt limits

webvictim committed:
S3 folder names aren't case-sensitive

webvictim committed:
Change directory

webvictim committed:
Selective build

webvictim committed:
Updates

webvictim committed:
Fix up Jenkinsfile

webvictim committed:
Add error check to S3 bucket deletion

webvictim committed:
Don't try and override bucket ID in Makefile

webvictim committed:
Merge branch 'master' into gus/cf-existing-vpc-support

webvictim committed:
Update Teleport to 2.6.0

webvictim committed:
Force deregister Packer snapshots

webvictim committed:
Merge branch 'gus/cf-existing-vpc-support' of https://github.com/gravitational/teleport into gus/cf-existing-vpc-support

webvictim committed:
Re-enable automatic VPC creation/deletion

webvictim commented:
retest this please

webvictim commented:
- Cloudformation now expects an existing VPC, 2 public and 4 private subnets to run
- Updated Teleport to version 2.6.0
- Randomised domain name to make SSL certificate issuing work more frequently
- S3 buckets are now deleted after running
- Old AMIs/snapshots are deleted automtaically

webvictim commented:
Fixes #1973 
Fixes #1979 
Fixes #1982
